synapse/media/media_repository.py

```diff
@@ -276,6 +277,14 @@ async def get_remote_media(
         ):
             raise FederationDeniedError(server_name)
 
+        # Don't let users download media from domains listed in the config, even
+        # if we might have the media to serve. This is Trust & Safety tooling to
+        # block some servers' media from being accessible to local users.
+        # See `prevent_downloads_from` config docs for more info.
+        if server_name in self.prevent_downloads_from:
+            respond_404(request)
```

Comment:
Neither `404` nor `403` are valid return codes in the [spec](https://spec.matrix.org/latest/client-server-api/#get_matrixmediav3downloadservernamemediaid). This is either intentional or a bug.
But IMO here is `403 forbidden` the better choice. It is forbidden here for security reasons. A `not found` would be a pseudo `not found` and rather misleading when searching for errors.

Reply:
The spec is forgetting error codes for sure. 

It's questionable whether we'd want to reveal the reason why to callers though. I'm also hesitant to use auth errors as that can cause some clients to log the user out.

Reply:
> It's questionable whether we'd want to reveal the reason why to callers though. I'm also hesitant to use auth errors as that can cause some clients to log the user out.

That is not nice.

> However, a request might be forbidden for reasons unrelated to the credentials.

https://www.rfc-editor.org/rfc/rfc9110.html#name-403-forbidden

Reply:
indeed - it comes from our use of error codes to communicate intent, and historically clients took it a step further and translated that to http error codes.

Reply:
fwiw, on my own project it comes through primarily as a 404, but with a clearer message and sub-errorcode to clarify intent: https://github.com/turt2live/matrix-media-repo/commit/219e4bc67ab8e354b263774aca0118332ea48c25

Reply:
I'd go for `403 Forbidden` and I don't see why a client should see that as a logout request. If we were using `401 Unauthorized`, then I can see a server eventually trigger a re-authentication request.

Reply:
Will leave this up to the Synapse team to decide.